internal/integration/ogent/ent/ogent/ogent.go

```diff
@@ -29,6 +34,245 @@ func rawError(err error) jx.Raw {
+}
+
+type SymbolRef struct {
+	Symbol     string        `@Ident @{ "." Ident }`
+	Parameters []*Expression `[ "(" @@ { "," @@ } ")" ]`
+}
+
+type Value struct {
+	Wildcard bool       `(  @"*"`
+	Number   *float64   ` | @Number`
+	String   *string    ` | @String`
+	Boolean  *Boolean   ` | @("TRUE" | "FALSE")`
+	Null     bool       ` | @"NULL"`
+	Date     *time.Time ` | @Date`
+	Array    *Array     ` | @@ )`
+}
+
+type Array struct {
+	Expressions []*Expression `"(" @@ { "," @@ } ")"`
+}
+
+func ParseExpression(exp *Expression, s *sql.Selector) {
+	i := 0
+	len := len(exp.Or)
+	for i < len { //OR
+		parseAndExpression(exp.Or[i], s)
+		i++
+		if i < len {
+			s.Or()
+		}
+	}
+}
+
+func parseAndExpression(and *OrCondition, s *sql.Selector) {
+	for _, c := range and.And {
+		if c.Operand != nil {
+			addPredicate(c, s)
+		} else if c.Not != nil {
+			addPredicate(c.Not, s)
+		}
+	}
+}
+
+func addPredicate(c *Condition, s *sql.Selector) {
+	con := c.Operand.ConditionRHS
+	if con != nil {
+		op := con.Compare.Operator
+		col := getColumn(c.Operand)
+		v := getOperand(c.Operand.ConditionRHS.Compare.Operand, s)
+		switch op {
+		case "eq":
+			s.Where(sql.EQ(s.C(col), v)) //Equal
+		case "ne":
+			s.Where(sql.NEQ(s.C(col), v)) //Not equal
+		case "gt":
+			s.Where(sql.GT(s.C(col), v)) //Greater than
+		case "ge":
+			s.Where(sql.GTE(s.C(col), v)) //Greater than or equal
+		case "lt":
+			s.Where(sql.LT(s.C(col), v)) //Less than
+		case "le":
+			s.Where(sql.LTE(s.C(col), v)) //Less than or equal
+		}
+	} else {
+		v := getOperand(c.Operand.Operand, s).(*Array)
+		for _, i := range v.Expressions {
+			ParseExpression(i, s)
+		}
+	}
+}
+
+func getColumn(col *ConditionOperand) string {
+	return getOperand(col.Operand, nil).(string)
+}
+
+func getOperand(op *Operand, s *sql.Selector) interface{} {
+	for _, su := range op.Summand {
+		return getValue(su, s)
+	}
+	return nil
+}
+
+func getValue(su *Summand, s *sql.Selector) interface{} {
+	if su.LHS.LHS.Value != nil {
+		v := su.LHS.LHS.Value
+		if v.String != nil {
+			return v.String
+		} else if v.Number != nil {
+			return v.Number
+		} else if v.Date != nil {
+			return v.Date
+		} else if v.Array != nil {
+			return v.Array
+		}
+	} else {
+		return su.LHS.LHS.SymbolRef.Symbol
+	}
+	return nil
+}
+
+var (
+	sqlLexer = lexer.Must(lexer.Regexp(`(\s+)` +
+		`|(?P<Keyword>(?i)SELECT|FROM|TOP|DISTINCT|ALL|WHERE|GROUP|BY|HAVING|UNION|MINUS|EXCEPT|INTERSECT|ORDER|LIMIT|OFFSET|TRUE|FALSE|NULL|IS|NOT|ANY|SOME|BETWEEN|and|or|LIKE|AS|IN)` +
+		`|(?P<Ident>[a-zA-Z_][a-zA-Z0-9_]*)` +
+		`|(?P<Date>\d{4})-(\d{2})-(\d{2})[tT](\d{2}):(\d{2}):(\d{2})(\.\d+)?([zZ]|(\+|-)(\d{2}):(\d{2}))` +
+		`|(?P<Number>[-+]?\d*\.?\d+([eE][-+]?\d+)?)` +
+		`|(?P<String>'[^']*'|"[^"]*")` +
+		`|(?P<Operators><>|!=|<=|>=|[-+*/%,.()=<>]|eq|ne|gt|ge|lt|le)`,
+	))
+	SqlParser = participle.MustBuild(
+		&Expression{},
+		participle.Lexer(sqlLexer),
+		participle.Unquote("String"),
+		participle.CaseInsensitive("Keyword"),
+	)
```

Comment:
Why do we need a parser here, for the filter predicates? Similar to the [`WhereInput`](https://entgo.io/docs/tutorial-todo-gql-filter-input) objects we have in entgql, I think we can use the same approach here. See the template for it [here](https://github.com/ent/contrib/blob/master/entgql/template/where_input.tmpl).

In order to keep this API secure, filters should be translated to ent predicates, that are predefined in the generated packages.

Reply:
I understand the scenario of having the API secure. The code created by the ent predefining the predicates works very well. However, in the case of tables with hundred columns, the performance could be degraded.
My current project for using ogent is to have over a hundred tables with a hundred columns each. Generating thousands of API calls. By creating this generic way of adding the predicates I will gain good performance. I updated the code validating if the column added inside the filter is valid or not.

Reply:
> I understand the scenario of having the API secure. The code created by the ent predefining the predicates works very well. However, in the case of tables with hundred columns, the performance could be degraded. My current project for using ogent is to have over a hundred tables with a hundred columns each. Generating thousands of API calls. By creating this generic way of adding the predicates I will gain good performance. I updated the code validating if the column added inside the filter is valid or not.

I don't see why having a generated (specific) code part for the hundreds of columns is degrading performance compared to a generic approach. Quite the contrary should be the case, because there should be fewer cycles to run through with a well defined code path. Am I missing something?

Reply:
I checked the GQL code and for each field, it creates 7 variables, one for each predicate "field-name+predicate". As I mentioned before, we have 100 tables with 100 columns. It means the GQL will create 700 if cases for each table to check one by one what predicate will be added in my query and it will always happen even if I have only one single column in my filter. However, the generic approach will have only one single loop if my filter contains a single column like "name eg ''Bob".